pkg/api/validation/validation_test.go

```diff
@@ -2144,172 +2178,85 @@ func TestValidateMinionUpdate(t *testing.T) {
-					Labels: map[string]string{"foo": "bar"},
-				},
+		{
+			name: "no change",
+			tweakSvc: func(oldSvc, newSvc *api.Service) {
+				// do nothing
 			},
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:   "foo",
-					Labels: map[string]string{"foo": "baz"},
-				},
-			}, true},
-		{ // 3
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name: "foo",
-				},
+			numErrs: 0,
+		},
+		{
+			name: "change name",
+			tweakSvc: func(oldSvc, newSvc *api.Service) {
+				newSvc.Name += "2"
 			},
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:   "foo",
-					Labels: map[string]string{"foo": "baz"},
-				},
-			}, true},
-		{ // 4
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:   "foo",
-					Labels: map[string]string{"bar": "foo"},
-				},
+			numErrs: 1,
+		},
+		{
+			name: "change namespace",
+			tweakSvc: func(oldSvc, newSvc *api.Service) {
+				newSvc.Namespace += "2"
 			},
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:   "foo",
-					Labels: map[string]string{"foo": "baz"},
-				},
-			}, true},
-		{ // 5
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:        "foo",
-					Annotations: map[string]string{"bar": "foo"},
-				},
+			numErrs: 1,
+		},
+		{
+			name: "change label valid",
+			tweakSvc: func(oldSvc, newSvc *api.Service) {
+				newSvc.Labels["key"] = "other-value"
 			},
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:        "foo",
-					Annotations: map[string]string{"foo": "baz"},
-				},
-			}, true},
-		{ // 6
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name: "foo",
-				},
-				Spec: api.ServiceSpec{
-					Selector: map[string]string{"foo": "baz"},
-				},
+			numErrs: 0,
+		},
+		{
+			name: "add label",
+			tweakSvc: func(oldSvc, newSvc *api.Service) {
+				newSvc.Labels["key2"] = "value2"
 			},
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name: "foo",
-				},
-				Spec: api.ServiceSpec{
-					Selector: map[string]string{"foo": "baz"},
-				},
-			}, true},
-		{ // 7
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:   "foo",
-					Labels: map[string]string{"bar": "foo"},
-				},
-				Spec: api.ServiceSpec{
-					PortalIP: "127.0.0.1",
-				},
+			numErrs: 0,
+		},
+		{
+			name: "change portal IP",
+			tweakSvc: func(oldSvc, newSvc *api.Service) {
+				oldSvc.Spec.PortalIP = "1.2.3.4"
+				newSvc.Spec.PortalIP = "8.6.7.5"
 			},
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:   "foo",
-					Labels: map[string]string{"bar": "fooobaz"},
-				},
-				Spec: api.ServiceSpec{
-					PortalIP: "new",
-				},
-			}, false},
-		{ // 8
-			api.Service{
-				ObjectMeta: api.ObjectMeta{
-					Name:   "foo",
-					Labels: map[string]string{"bar": "foo"},
-				},
-				Spec: api.ServiceSpec{
-					PortalIP: "127.0.0.1",
-				},
+			numErrs: 1,
+		},
+		{
+			name: "remove portal IP",
+			tweakSvc: func(oldSvc, newSvc *api.Service) {
+				oldSvc.Spec.PortalIP = "1.2.3.4"
+				newSvc.Spec.PortalIP = ""
```

Comment:
Hi Tim, I have a issue #5907 related to this part. `newSvc.Spec.PortalIP = ""` usually happens when portalIP is not given in the template. Since when creating a service not giving a portalIP means allow kubernetes automatically allocate one, I think we should keep this semantic while updating. That is use the system allocated one. For removing portalIP we can use something like `portalIP: None` explicitly in the template. What do you think?